pkg/collect/utilisation.go

```diff
@@ -0,0 +1,363 @@
+	}
+
+	return prometheusClient, stopCh, err
+}
+
+type PortForwardFn = func(*rest.Config, string, string, int, output.Output) (int, chan struct{}, error)
+
+// PortForward creates an asynchronous port-forward to a pod on the cluster in order to
+// access it over the local network.
+func PortForward(cfg *rest.Config, ns, pod string, port int, out output.Output) (localPort int, stop chan struct{}, err error) {
+	dialer, err := getDialer(cfg, ns, pod)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	// configure the portforwarder
+	stop, ready := make(chan struct{}, 1), make(chan struct{}, 1)
+	pf, localPort, err := getPortForwarder(dialer, port, stop, ready, out)
+	if err != nil {
+		return localPort, nil, err
+	}
+
+	if err := runPortForwarder(pf, ready, out); err != nil {
+		return -1, nil, err
+	}
+
+	return localPort, stop, nil
+}
+
+func getDialer(cfg *rest.Config, ns, pod string) (httpstream.Dialer, error) {
+	roundTripper, upgrader, err := spdy.RoundTripperFor(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	serverURL, err := constructPortForwardURL(cfg.Host, ns, pod)
+	if err != nil {
+		return nil, err
+	}
+
+	return spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, serverURL), nil
+}
+
+func getPortForwarder(dialer httpstream.Dialer,
+	port int, stop, ready chan struct{}, out output.Output,
+) (*portforward.PortForwarder, int, error) {
+	// get an available local port to use
+	localPort, err := freeport.GetFreePort()
+	if err != nil {
+		return nil, localPort, err
+	}
+
+	// configure the portforwarder
+	pf, err := portforward.New(dialer,
+		[]string{fmt.Sprintf("%d:%d", localPort, port)}, stop, ready, out.InfoWriter(), out.ErrorWriter())
+
+	return pf, localPort, err
+}
+
+func runPortForwarder(pf *portforward.PortForwarder, ready chan struct{}, out output.Output) error {
+	out.Info("running the port-forwarder in the background")
+	errCh := make(chan error)
+	// start the portforwarder in the background and look for errors
+	go func() {
+		if err := pf.ForwardPorts(); err != nil {
+			errCh <- err
+		}
+		close(errCh)
+	}()
+	// wait for the port-forward to be established and check for errors
+	out.Info("waiting for port-forward to be established")
+	select {
+	case <-ready:
+		out.Info("the port-forward has signaled ready")
+	case err := <-errCh:
+		if err != nil {
+			return fmt.Errorf("port-forward failed: %w", err)
+		}
+	}
+	return nil
+}
+
+func constructPortForwardURL(host, ns, pod string) (*url.URL, error) {
+	hostURL, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+
+	return &url.URL{
+		Scheme: "https",
+		Path: path.Join(
+			hostURL.Path,
+			"api/v1/namespaces",
+			url.PathEscape(ns),
+			"pods",
+			url.PathEscape(pod),
+			"portforward",
+		),
+		Host: hostURL.Host,
+	}, nil
+}
+
+func joinReadings(cpuSamples, memSamples []model.SamplePair) ([]UsageReading, error) {
+	jointReadingsList := make([]UsageReading, 0)
```

Comment:
the length is at least the minimum of len(cpuSamples) and len(memSamples) right?